src/Microsoft.IdentityModel.Tokens/EventBasedLRUCache.cs

```diff
@@ -336,69 +305,102 @@ public bool TryRemove(TKey key, out TValue value)
+            if (_removeExpiredValues && _taskCount == 0) // pause the timer only if the _taskCount is 0 to avoid the scenario that it is being resumed (above)
             {
-                _eventQueueTaskState = EventQueueTaskRunning; // set to the state back to running
+                PauseTimer();
             }
         }
 
         /// <summary>
-        /// The method handling the event when an item is added to the LinkedList.
-        /// The _eventQueueTask needs to be started if the _doubleLinkedList is empty.
-        /// 
-        /// This method is called after an item is added to the queue, so it needs to start the event queue task if it is not running (_eventQueueTaskState == EventQueueTaskRunning),
-        /// and prevent the task from being stopped if it has been set to stop but not exited yet (the task while loop is still running).
+        /// This method is called after an item is added to the event queue, and the event queue task needs to be started if not running (_eventQueueTaskState != EventQueueTaskRunning).
+        /// Settitng _eventQueueTaskState to EventQueueTaskRunning should prevent multiple task being started.
         /// </summary>
-        private void StartEventQueueTasksIfNotRunning()
+        private void StartEventQueueTaskIfNotRunning()
         {
-            // Setting _eventQueueTaskState to EventQueueTaskRunning will keep the event queue task in EventQueueTaskAction running.
-            if (Interlocked.CompareExchange(ref _eventQueueTaskState, EventQueueTaskRunning, EventQueueTaskStopRequested) == EventQueueTaskStopRequested)
+             _eventQueueTaskStopTime = SetTaskEndTime(); // set the time when the _eventQueueTask should end
+
+            // Setting _eventQueueTaskState to EventQueueTaskDoNotStop here will force the event queue task in EventQueueTaskAction to continue even it has past the _eventQueueTaskEndTime.
+            // It is mainly to prevent a rare (but theoratically possible) scenario caused by context switching
+            // For example:
+            //   1. the task execution in EventQueueTaskAction() checks _eventQueueTaskStopTime and it has already passed (ready to exit)
+            //   2. the execution is switched to this thread (before it calls the Interlocked.CompareExchange() to set the _eventQueueTaskState to EventQueueTaskStopped)
+            //   3. the _eventQueueTaskStopTime is extended but it can't stop the task from stopping as the task has already passed the time check
+            //   4. now since the _eventQueueTaskState == EventQueueTaskRunning, it can be set to EventQueueTaskDoNotStop by the Interlocked.CompareExchange() below
+            //   5. if _eventQueueTaskState is successfully set to EventQueueTaskDoNotStop, the Interlocked.CompareExchang() in the EventQueueTaskAction() will fail
+            //      and the task will continue the while loop and the new _eventQueueTaskStopTime will keep the task running
```

Comment:
can we also add this comment on line 339

Reply:
Added more comments hopefully will help anyone to better understand the logic, thanks.

Reply:
More comments added to explain the logic, thanks.